parsers/UY.py

```diff
@@ -1,163 +1,264 @@
 #!/usr/bin/python3
-
 import re
 from datetime import datetime
 from logging import Logger, getLogger
-from typing import Optional
+from typing import Callable, List, Optional
 
 import arrow
 import dateutil
-
-# BeautifulSoup is used to parse HTML to get information
 from bs4 import BeautifulSoup
 from requests import Session
 
-tz = "America/Montevideo"
+from parsers.lib.exceptions import ParserException
 
+TIME_ZONE = "America/Montevideo"
+
+# maps the xml keys from the website to names used by parser system
 MAP_GENERATION = {
-    "Hidráulica": "hydro",
-    "Eólica": "wind",
-    "Fotovoltaica": "solar",
-    "Biomasa": "biomass",
-    "Térmica": "oil",
+    # short for rio negro, main hydro power plant
+    "hydro": "r",
+    "wind": "eolica",
+    "solar": "fotovoltaica",
+    "biomass": "biomasa",
+    # all thermal, i.e. miscilaneous oil + natural gas production
+    "unknown": "termica",
+    "trade": "intercambios",
+    "demand": "demanda",
+    "salto_grande_prod": "comprassgu",
 }
-INV_MAP_GENERATION = dict([(v, k) for (k, v) in MAP_GENERATION.items()])
+
+AVALIABLE_KEYS = ["hydro", "wind", "solar", "biomass", "unknown"]
+
+UTE_URL = url = "https://ute.com.uy/energia-generada-intercambios-demanda"
 
 SALTO_GRANDE_URL = "http://www.cammesa.com/uflujpot.nsf/FlujoW?OpenAgent&Tensiones y Flujos de Potencia&"
 
 
-def get_salto_grande(session: Optional[Session]) -> float:
+def get_salto_grande(session: Session, targ_time: datetime = None) -> float:
     """Finds the current generation from the Salto Grande Dam that is allocated to Uruguay."""
 
+    # Data for current hour seems to be available after 30mins.
+    # if we're too soon into the hour, check the hour before
     current_time = arrow.now("UTC-3")
-    if current_time.minute < 30:
-        # Data for current hour seems to be available after 30mins.
-        current_time = current_time.shift(hours=-1)
-    lookup_time = current_time.floor("hour").format("DD/MM/YYYY HH:mm")
+    if (
+        current_time.floor("hour") == targ_time.floor("hour")
+        and current_time.minute < 30
+    ):
+        # Looking at previous hour's data for salto grande, because it is too soon after the hour
+        targ_time = targ_time.shift(hours=-1)
```

Comment:
Isn't there a risk we get data for the previous hour of salto grande and the current hour for the test with this?

You probably need to add some validation somewhere to make sure both datetimes match.

Reply:
this comparison is more or less verbatim what was done from commit [63bd25e](https://github.com/electricitymaps/electricitymaps-contrib/pull/1334/commits/63bd25e234f491bcf65107c04beef034295eecbc) when this method was introduced.

Reply:
Hmm looking at that presents another problem, do you know if the dates used is forward looking or back looking? i.e. does 10:00 include the production between 10:00 and 11:00 or between 9:00 and 10:00?

Electricity Maps use the first and I suspect this parser use the second so we might have to shift the datetimes around.

In either case the problem with this function in particular is that if the time is say 10:31 salto grande would show the production for 10:00 while the other source might show the production for 9:00. We would need to compare the datetimes in each response to make sure they are reporting production for the same hour so we don't get a mismatch between sources. This should have been done in the previous version of parser as well.